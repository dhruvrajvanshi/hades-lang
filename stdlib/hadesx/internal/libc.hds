type int = i32

extern fn fork(): int = fork
extern fn wait(int): Void = wait
