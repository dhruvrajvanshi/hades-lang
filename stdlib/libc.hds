// TODO: do something better here. This isn't portable.
type CInt = i32;
type int = i32;
type uint = u32
type ulong = u64
type long = i64

extern def puts(*Byte): Void = puts;
extern def exit(CInt): Void = exit;
extern def malloc(Size): *mut Void = malloc;
extern def free(*mut Void): Void = free;
extern def strlen(*Byte): usize = strlen;
extern def strcat(*Byte, *Byte): *mut Byte = strcat;
extern def strcmp(*Byte, *Byte): CInt = strcmp;
extern def strcpy(*mut Byte, *Byte): *mut Byte = strcpy;
extern def memcpy(*Void, *Void, usize): Void = memcpy;
extern def memcmp(*Void, *Void, usize): CInt = memcmp;
extern def floor(Double): Double = floor;

extern def fprintf(*File, *Byte, *Byte): int = fprintf;
extern def printf(*Byte, *Byte): int = printf;
extern def sleep(uint): uint = sleep;
extern def fflush(*File): Void = fflush;

extern def fputs(*Byte, *File): Void = fputs;

struct File {}

extern def getline(*mut *mut Byte, *mut usize, *File): isize = getline;
extern def fgetc(*File): Byte = fgetc;
extern def strtoul(*Byte, **Byte, u32): ulong = strtoul;


extern def fread(*Void, usize, usize, *File): usize = fread;

type FileDescriptor = int;
extern def read(FileDescriptor, *mut Void, usize): isize = read;
extern def write(FileDescriptor, *Void, usize): isize = write;

extern def errno(): int = hdc_errno;
extern def strerror(CInt): *Byte = strerror;
extern def set_errno(CInt): Void = hdc_set_errno;

extern def open(*Byte, CInt): CInt = open;

extern def stdout(): *File = hdc_get_stdout;
extern const stdin: *File = stdin;
extern def stderr(): *File = hdc_get_stderr;

extern def perror(*Byte): Void = perror;

extern def listen(CInt, CInt): CInt = listen;

struct sockaddr {}
struct socklen_t {}
extern def accept(CInt, *mut sockaddr, *mut socklen_t): CInt = accept;
extern def close(CInt): CInt = close;


const EOF: int = -1

const SEEK_SET: int = 0
const SEEK_CUR: int = 1
const SEEK_END: int = 2

/// fseek(file, offset, whence)
/// whence should be SEEK_SET or SEEK_CUR or SEEK_END
extern def fseek(*File, long, int): i32 = fseek

extern def ftell(*File): long = ftell

/// fopen(file, mode)
extern def fopen(*Byte, *Byte): *File = fopen
extern def fclose(*File): int = fclose
extern def rewind(*File): Void = rewind
extern def feof(*File): int = feof
extern def ferror(*File): int = ferror
extern def fileno(*File): int = fileno
extern def abort(): Void = abort

type pid_t = int
extern def fork(): pid_t = fork

/// https://man7.org/linux/man-pages/man2/waitpid.2.html
extern def waitpid(pid_t, *int, int): pid_t = waitpid
extern def wait(*int): pid_t = wait
