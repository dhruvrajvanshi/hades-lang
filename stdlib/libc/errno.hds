type t = Int;

fn EPERM(): Int { return  1; }
fn ENOENT(): Int { return  2; }
fn ESRCH(): Int { return  3; }
fn EINTR(): Int { return  4; }
fn EIO(): Int { return  5; }
fn ENXIO(): Int { return  6; }
fn E2BIG(): Int { return  7; }
fn ENOEXEC(): Int { return  8; }
fn EBADF(): Int { return  9; }
fn ECHILD(): Int { return 10; }
fn EAGAIN(): Int { return 11; }
fn ENOMEM(): Int { return 12; }
fn EACCES(): Int { return 13; }
fn EFAULT(): Int { return 14; }
fn ENOTBLK(): Int { return 15; }
fn EBUSY(): Int { return 16; }
fn EEXIST(): Int { return 17; }
fn EXDEV(): Int { return 18; }
fn ENODEV(): Int { return 19; }
fn ENOTDIR(): Int { return 20; }
fn EISDIR(): Int { return 21; }
fn EINVAL(): Int { return 22; }
fn ENFILE(): Int { return 23; }
fn EMFILE(): Int { return 24; }
fn ENOTTY(): Int { return 25; }
fn ETXTBSY(): Int { return 26; }
fn EFBIG(): Int { return 27; }
fn ENOSPC(): Int { return 28; }
fn ESPIPE(): Int { return 29; }
fn EROFS(): Int { return 30; }
fn EMLINK(): Int { return 31; }
fn EPIPE(): Int { return 32; }
fn EDOM(): Int { return 33; }
fn ERANGE(): Int { return 34; }

fn description(err: Int): *Byte {
    if err == EPERM() {
        return c"EPERM: Operation not permitted";
    }
    if err == ENOENT() {
        return c"ENOENT: No such file or directory";
    }
    if err == ESRCH() {
        return c"ESRCH: No such process";
    }
    if err == EINTR() {
        return c"EINTR: Interrupted system call";
    }
    if err == EIO() {
        return c"EIO: I/O error";
    }
    if err == ENXIO() {
        return c"ENXIO: No such device or address";
    }
    if err == E2BIG() {
        return c"E2BIG: Argument list too long";
    }
    if err == ENOEXEC() {
        return c"ENOEXEC: Exec format error";
    }
    if err == EBADF() {
        return c"EBADF: Bad file number";
    }
    if err == ECHILD() {
        return c"ECHILD: No child processes";
    }
    if err == EAGAIN() {
        return c"EAGAIN: Try again";
    }
    if err == ENOMEM() {
        return c"ENOMEM: Out of memory";
    }
    if err == EACCES() {
        return c"EACCES: Permission denied";
    }
    if err == EFAULT() {
        return c"EFAULT: Bad address";
    }
    if err == ENOTBLK() {
        return c"ENOTBLK: Block device required";
    }
    if err == EBUSY() {
        return c"EBUSY: Device or resource busy";
    }
    if err == EEXIST() {
        return c"EEXIST: File exists";
    }
    if err == EXDEV() {
        return c"EXDEV: Cross-device link";
    }
    if err == ENODEV() {
        return c"ENODEV: No such device";
    }
    if err == ENOTDIR() {
        return c"ENOTDIR: Not a directory";
    }
    if err == EISDIR() {
        return c"EISDIR: Is a directory";
    }
    if err == EINVAL() {
        return c"EINVAL: Invalid argument";
    }
    if err == ENFILE() {
        return c"ENFILE: File table overflow";
    }
    if err == EMFILE() {
        return c"EMFILE: Too many open files";
    }
    if err == ENOTTY() {
        return c"ENOTTY: Not a typewriter";
    }
    if err == ETXTBSY() {
        return c"ETXTBSY: Text file busy";
    }
    if err == EFBIG() {
        return c"EFBIG: File too large";
    }
    if err == ENOSPC() {
        return c"ENOSPC: No space left on device";
    }
    if err == ESPIPE() {
        return c"ESPIPE: Illegal seek";
    }
    if err == EROFS() {
        return c"EROFS: Read-only file system";
    }
    if err == EMLINK() {
        return c"EMLINK: Too many links";
    }
    if err == EPIPE() {
        return c"EPIPE: Broken pipe";
    }
    if err == EDOM() {
        return c"EDOM: Math argument out of domain of func";
    }
    if err == ERANGE() {
        return c"ERANGE: Math result not representable";
    }
    return c"";
}
