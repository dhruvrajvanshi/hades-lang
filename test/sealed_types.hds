import submodule.printable_interface as p;
import libc as c;

def main(): Void {
    val opt = Optional.Some(10);
    get_int_value(opt).print();
    get_int_value(Optional.Some(20)).print();
    get_int_value(Optional.None[CInt]).print();

    val opt_message = Optional.Some[*Byte](b"message");
    val opt_message_none = Optional.None[*Byte];

    c.puts(get_string_value(opt_message));
    c.puts(get_string_value(opt_message_none));
}

def get_string_value(value: Optional[*Byte]): *Byte {
    return when value {
        is s: Some -> message,
        is None -> b"None"
    };
}

def get_int_value(value: Optional[CInt]): CInt {
    return when value {
        is s: Some -> s.value,
        is None -> 0
    };
}

sealed type Optional[T] {
    Some(value: T);
    None;
}

