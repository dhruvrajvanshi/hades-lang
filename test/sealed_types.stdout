10
20
0
message
None
